File: LeetCode/Easy/0338-counting-bits/0338-counting-bits.java
class Solution {
    public int[] countBits(int n) {
        int[] ret = new int[n + 1];
        if (n == 0) return ret;
        ++ret[1];
        if (n == 1) return ret;
        int i = 2;
        ++n;
        while (n > i){
            if (n >= i * 2){
                for (int j = 0; j < i; ++j) ret[i+j] = ret[j] + 1;
                i *= 2;
            } else {
                for (int j = 0; j < n - i; ++j) ret[i+j] = ret[j] + 1;
                i *= 2;
            }
        }
        return ret;
    }
}

File: LeetCode/Easy/0338-counting-bits/0338-counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        if n == 0:
            return [0]
        if n == 1:
            return [0, 1]
        ret = [0, 1]
        i = 2
        n += 1
        while n > i:
            if n >= i * 2:
                ret += [i + 1 for i in ret]
                i *= 2
            else:
                ret += [ret[i] + 1 for i in range(n - i)]
                i *= 2
        return ret
            
            

